### PR TITLE
Flush immediately after a remote recovery finishes (unless there are ongoing ones)

### PR DESCRIPTION
**Pull request by bleskes**
To properly replicate, we currently stop flushing during recovery so we can repay the translog once copying files are done. Once recovery is done, the translog will be flushed by a background thread that, by default, kicks in every 5s. In case of a recovery failure and a quick re-assignment of a new shard copy, we may fail to flush before starting a new recovery, causing it to deal with potentially even longer translog. This commit makes sure we flush immediately when the ongoing recovery count goes to 0.

I also added a simple recovery benchmark.

### CODE REVIEW

File: src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java [hunk 1]
```diff
@@ -1681,8 +1681,20 @@ public int get() {
 
         public void endRecovery() throws ElasticsearchException {
             store.decRef();
-            onGoingRecoveries.decrementAndGet();
+            int left = onGoingRecoveries.decrementAndGet();
             assert onGoingRecoveries.get() >= 0 : "ongoingRecoveries must be >= 0 but was: " + onGoingRecoveries.get();
+            if (left == 0) {
```
Comment (mikemccand) [inline src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java]: I really don't understand the concurrency here, but I just wanted to verify that there is no risk that another recovery kicks off right here, after a first recovery just finished and dropped the count to 0, and then we clear the xlog out from under that second recovery, losing documents?

**Comment by mikemccand** [comment]
LGTM, but I left one concurrency question...

File: src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java (hunk 1 above)
Comment (bleskes) [inline src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java]: this is protected in the flush logic it self, so we can safely try:

https://github.com/elasticsearch/elasticsearch/blob/1.4/src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java#L841

https://github.com/elasticsearch/elasticsearch/blob/1.4/src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java#L855

and on the recovery side:

https://github.com/elasticsearch/elasticsearch/blob/1.4/src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java#L1125

Comment (mikemccand) [inline src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java]: OK, I see, looks good.  Thanks @bleskes

**Comment by imotov** [comment]
LGTM

File: src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java [hunk 2]
```diff
@@ -1681,8 +1681,19 @@ public int get() {
...
+                try {
+                    flush(new Engine.Flush().type(Flush.Type.COMMIT_TRANSLOG));
```
Comment (dakrone) [inline src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java]: For the 1.x and master future-port, how are you going to implement the flush? Are you going to set `force` and `waitIfOngoing`?

**Comment by dakrone** [comment]
LGTM, left one comment

File: src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java (hunk 2 above)
Comment (bleskes) [inline src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java]: I think both can be left at their defaults? `force=false` `waitIfOngoing=false`. It means there is already an ongoing flush or there is nothing to do. In both cases, no need for an extra flush to reduce the translog.

Comment (dakrone) [inline src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java]: sounds good to me

**Pull request closed without merging**